xarray/core/coordinates.py

```diff
@@ -236,13 +449,86 @@ def merge(self, other: Coordinates | None) -> Dataset:
             variables=coords, coord_names=coord_names, indexes=indexes
         )
 
+    def __setitem__(self, key: Hashable, value: Any) -> None:
+        self.update({key: value})
+
+    def update(self, other: Mapping[Any, Any]) -> None:
+        """Update this Coordinates variables with other coordinate variables."""
+        other_obj: Coordinates | Mapping[Hashable, Variable]
+
+        if isinstance(other, Coordinates):
+            # special case: default indexes won't be created
+            other_obj = other
+        else:
+            other_obj = getattr(other, "variables", other)
+
+        self._maybe_drop_multiindex_coords(set(other_obj))
+
+        coords, indexes = merge_coords(
+            [self.variables, other_obj],
+            priority_arg=1,
+            indexes=self.xindexes,
+        )
+
+        self._update_coords(coords, indexes)
+
+    def _overwrite_indexes(
+        self: T_Coordinates,
+        indexes: Mapping[Any, Index],
+        variables: Mapping[Any, Variable] | None = None,
+    ) -> T_Coordinates:
+        results = self.to_dataset()._overwrite_indexes(indexes, variables)
+        return cast(T_Coordinates, results.coords)
+
+    def _reindex_callback(
+        self: T_Coordinates,
+        aligner: Aligner,
+        dim_pos_indexers: dict[Hashable, Any],
+        variables: dict[Hashable, Variable],
+        indexes: dict[Hashable, Index],
+        fill_value: Any,
+        exclude_dims: frozenset[Hashable],
+        exclude_vars: frozenset[Hashable],
+    ) -> T_Coordinates:
+        """Callback called from ``Aligner`` to create a new reindexed Coordinates."""
+        aligned = self.to_dataset()._reindex_callback(
+            aligner,
+            dim_pos_indexers,
+            variables,
+            indexes,
+            fill_value,
+            exclude_dims,
+            exclude_vars,
+        )
+        return cast(T_Coordinates, aligned.coords)
```

Comment:
I have not checked this, but maybe mypy is right to complain?

If you start with a DataArrayCoordinates, do you not end up with a DatasetCoordinates?

Reply:
> If you start with a DataArrayCoordinates, do you not end up with a DatasetCoordinates?

This case shouldn't happen because Dataset and DataArray have their own `_reindex_callback()` called internally in `Aligner`. The callback here is called when aligning a `Coordinates` object, so we need to cast `DatasetCoordinates` into a `Coordinates`.

I should probably add a comment here to explain this.